### PR TITLE
Revamp build system with ng-packagr

### PR DESCRIPTION
**Pull request by iam4x**
**Angular v6 introduces a schema to follow in order to build an distribute libraries.**

Since this is a hard task to always keep up to date I'm switching to [ng-packagr](https://github.com/dherges/ng-packagr) which is handling this task for us.

Warning, this is introducing breaking changes because we are not supporting Angular v4 anymore, this might be a no go but Angular v5 and v6 are now the way to go.

### CODE REVIEW

**Comment by iam4x** [comment]
Before we can test and merge this one, we need to solve: https://github.com/algolia/algoliasearch-client-javascript/pull/692

**Comment by Haroenv** [comment]
@iam4x I think you could either:

1. Look at other packages which have process.env in it and see what they do
2. Polyfill in this library as in the issue 
3. Use the dist/algoliasearchLite.min.js

**Comment by iam4x** [comment]
@Haroenv 

* Libraries having this issue they are fixing it inside the library itself, here iin our case it's algoliasearch-client-javascript

* You cannot polyfill the final application through a library, there is no way to interact with the global scope as we are in a module environment

* Did not know about `algoliasearchLite` will give it go 👍 

Thanks :)

**Comment by iam4x** [comment]
The algoliasearchLite uses `process.env` as well. We are pretty stuck here, we need to fix it in the original library :/

**Comment by vvo** [comment]
Can we move forward and document how users can solve this in their own apps for now? So that your nice PR can still be merged.

The Angular world is going in a way that is not one we planned for. When you are using React as a library, if process.env.NODE_ENV is development then it's in DEBUG mode. When using process.env.NODE_ENV as production then it's in production mode.

If Angular is not able to deal with that with libraries we need to find another way but it's not obvious how to do so. Maybe having algoliasearch.debug.js?

Again for now do you think we can move forward and document how to define process.env in your own app? Thanks!

**Comment by vvo** [comment]
cc @manekinekko any idea on how to build a library for Angular like the Algolia search client that:
- is not using process.env.NODE_ENV
- have the ability to have two different behaviours based on a global variable like REACT
- default to DEV mode if you don't provide another value?

Maybe that's not the right way to approach it, I dunno what's the right approach in the Angular world.

**Comment by iam4x** [comment]
I'm now running into a new issue with server side rendering and Angular v6: https://github.com/angular/angular/issues/23280

There's also a bug with a race condition with preboot https://github.com/angular/preboot/issues/82

We will have to wait form theses issues to be closed before upgrading the SSR example to Angular 6

(I'm reverting my work to Angular 5 for the SSR example)

**Comment by iam4x** [comment]
After some more investigation, it seems that the `routing: true` new option of InstantSearch.js is the culprit of the race condition with `preboot`.

We can't do much for it now, we will have to wait for the issue to be fixed inside `preboot`.

**Comment by iam4x** [comment]
Upgrade to Angular CLI v6 is done 🎉 

* Getting Started & Server Side Rendering guides have been updated
* Demos updated and running (https://deploy-preview-116--angular-instantsearch.netlify.com/examples/e-commerce/)

Now the build phase is entirely done with `ng-packagr`

**Comment by vvo** [comment]
giphy

File: .nvmrc [hunk 1]
```diff
@@ -1 +1 @@
-v9.5.0
+v9.11.1
```
Comment (vvo) [inline .nvmrc]: I guess we could go 10.x now that it's out

Comment (iam4x) [reply .nvmrc]: I've tried them all, we have restrictions from dependencies to run.

Actually:
	* examples/server-side-rendering: node8

File: community-website/src/community-project-boilerplate-docgen/src/guides/server-side-rendering.md [hunk 2]
```diff
@@ -9,10 +9,12 @@ This is an advanced guide, if you never used Angular InstantSearch, you should f
 
 You can find the result of this guide on the Angular InstantSearch [repository](https://github.com/algolia/angular-instantsearch/tree/master/examples/server-side-rendering).
 
-Angular InstantSearch is compatible with server-side rendering starting with Angular version 5. We provide an API that is easy to use with [@angular/universal](https://github.com/angular/universal) modules.
+Angular InstantSearch is compatible with server-side rendering with Angular version 5. We provide an API that is easy to use with [@angular/universal](https://github.com/angular/universal) modules.
 
 For simplicity we are going to use the [@angular/universal-starter](https://github.com/angular/universal-starter) boilerplate which is a minimal Angular starter for Universal JavaScript using TypeScript and Webpack.
 
+**Angular Universal is not working completely with version 6, you will need to checkout a specific commit of `universal-start` boilerplate.**
```
[the remark this answers is not in the document]
Comment (iam4x) [reply community-website/src/community-project-boilerplate-docgen/src/guides/server-side-rendering.md]: I prefer to link the issue into one of our issue to keep track, will be done with the ROADMAP 👍

**Comment by manekinekko** [comment]
awesome job @iam4x 🙌
@vvo sorry, i've been busy with work these days.

**Pull request merged**